storage/gcp/gcp.go

```diff
@@ -918,84 +1019,239 @@ func (d *dedupStorage) index(ctx context.Context, h []byte) (*uint64, error) {
 	}
 }
 
-// storeMappings stores the associations between the keys and IDs in a non-atomic fashion
-// (i.e. it does not store all or none in a transactional sense).
-//
-// Returns an error if one or more mappings cannot be stored.
-func (d *dedupStorage) storeMappings(ctx context.Context, entries []dedupeMapping) error {
-	m := make([]*spanner.MutationGroup, 0, len(entries))
-	for _, e := range entries {
-		m = append(m, &spanner.MutationGroup{
-			Mutations: []*spanner.Mutation{spanner.Insert("IDSeq", []string{"id", "h", "idx"}, []interface{}{0, e.ID, int64(e.Idx)})},
-		})
-	}
+// Decorator returns a function which will wrap an underlying Add delegate with
+// code to dedup against the stored data.
+func (d *DedupStorage) Decorator() func(f tessera.AddFn) tessera.AddFn {
+	return func(delegate tessera.AddFn) tessera.AddFn {
+		return func(ctx context.Context, e *tessera.Entry) tessera.IndexFuture {
+			if d.pushBack.Load() {
```

Comment:
As discussed in person, an alternative approach is to return the index if we can find a positive match. It's not clear which is the best; one may allow catchup to happen faster, but the other may help fend off duplicate writes that will otherwise be retried. A quick comment that outlines this, and that this might want to change would be helpful. As someone looking at this code fresh, I can't tell if this decision is load-bearing or open to change.

Reply:
Comments added here and on the var in the struct.